### DISCUSSION TITLE
How to upload addressed content to the DHT automatically based on Kademlia?

### DISCUSSION THREAD

**Discussion by snowmead** [libp2p/rust-libp2p#4909] [Q&A] [answered]
From file-sharing and distributed-key-value-store examples, we are explicitly providing a key or file for a specific node. But my presumptions on Kademlia is that it will handle which node should hold on to which content based on their address and how "close" they are to that node's address. 

What is a standard approach to accomplish this in libp2p if there is any?

Scenario would be:

User sends chunks of data to a random node in the DHT network -> node (kademlia under the hood) would send these chunks to other nodes for storage

**Comment by thomaseizinger** [comment, marked as answer]
A DHT is a _discovery_ protocol, not a storage protocol. Using provider records, you can figure out who is _meant to_ store (aka "provide") a piece of data.

Kademlia doesn't store data for you. You need a separate protocol to distribute the data in your network, for example bitswap.

**Reply by thomaseizinger** [reply]
The issue on adding a bitswap implementation is here: https://github.com/libp2p/rust-libp2p/issues/2632.

**Reply by thomaseizinger** [reply]
Note that you can also easily roll your own data exchange protocol based on request-response. It somewhat depends on what you are building :)

**Reply by thomaseizinger** [reply]
At the moment, there is no standard approach to do this.

**Reply by snowmead** [reply]
Thank you so much for the detailed response, this really cleared things up for me. As you may already tell, I am new to the world of p2p. :)

**Reply by thomaseizinger** [reply]
No worries at all :)
